src/MigrationLog.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace P4\MasterTheme;
+
+/**
+ * A log of all migrations that have run, saved as a WP option.
+ */
+class MigrationLog {
+	/**
+	 * The WP option key.
+	 */
+	private const OPTION_KEY = 'planet4_migrations';
+
+	/**
+	 * @var array[] An entry for each done migration.
+	 */
+	private $done_migrations;
+
+	/**
+	 * Get the log from the WP options.
+	 *
+	 * @return static The log.
+	 */
+	public static function from_wp_options(): self {
+		$done_migrations = get_option( self::OPTION_KEY, [] );
+
+		$log = new self();
+
+		$log->done_migrations = $done_migrations;
+
+		return $log;
+	}
+
+	/**
+	 * Check whether a migration has already run.
+	 *
+	 * @param string $migration_id The migration to check.
+	 *
+	 * @return bool Whether said migration already ran.
+	 */
+	public function already_ran( string $migration_id ): bool {
+		foreach ( $this->done_migrations as $migration ) {
+			if ( $migration['id'] === $migration_id ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Record a migration in the log.
+	 *
+	 * @param string $migration_id Id of the migration to log.
+	 */
+	public function add( string $migration_id ): void {
```

Comment:
We could achieve more separation of concerns if we create a dedicated `MigrationRecord` object that holds just the data that we want to persist when running the script, without also having to hold the logic of the script itself. Then we can provide that record as a parameter to the script. That way the classes holding the actual migration logic will be as simple as possible, i.e. they hold the function that will be run and nothing else, except for the record, but it needs that to record stuff. Tried that out here https://github.com/greenpeace/planet4-master-theme/pull/1172/files#diff-77810f38e224a79d69eaf60f418eb92eR17

Reply:
In the linked PR I didn't intentionally didn't change the signature of `already_ran`, since that string alone is sufficient to do that lookup. Otherwise `already_ran` would be tricky to use like it's used here https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/357/files#diff-e0afc9a2c1803d93ce60bc43e7b3d5faR232